### PR TITLE
bpo-38639: Optimize floor(), ceil() and trunc() for floats.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue38639

### CODE REVIEW

File: Modules/mathmodule.c [hunk 1]
```diff
@@ -1114,15 +1114,17 @@ math_ceil(PyObject *module, PyObject *number)
...
+            return result;
+        }
         if (PyErr_Occurred())
             return NULL;
-        return math_1_to_int(number, ceil, 0);
     }
-    result = _PyObject_CallNoArg(method);
-    Py_DECREF(method);
-    return result;
+    return math_1_to_int(number, ceil, 0);
```
Comment (mdickinson) [inline Modules/mathmodule.c]: I think we should be able to dispense with the complications in `math_1_to_int` and simply use `PyLong_FromDouble(ceil(number))` here (and similarly for `floor`, of course). `floor` and `ceil` should never be setting `errno` or raising any error conditions: there aren't any circumstances where they can produce a NaN from non-NaN input, or an non-finite result from finite input.

Comment (mdickinson) [reply Modules/mathmodule.c]: And of course there are badly-behaved libm implementations out there, but `floor` and `ceil` are quite hard to get wrong even for those.

Comment (mdickinson) [reply Modules/mathmodule.c]: > PyLong_FromDouble(ceil(number))

Gah. There should be a `PyFloat_AsDouble` in there somewhere, of course. But you know what I meant. :-)

**Review by mdickinson — APPROVED** [review]
Proposed changes LGTM. For things of exact type `float` (and possibly also for `float` subclasses), I think we could dispense with the `math_1_to_int` complications to get even more speedup. But that doesn't have to happen in this PR.

Do we have tests for the various cases of floats, float subclasses that don't override `__floor__`, float subclasses that *do* override `__floor__`, things that aren't floats but provide a `__float__` method, etc.? If not, would it be worth adding such tests?

**Comment by mdickinson** [comment]
> Do we have tests for the various cases ...

It occurs to me that I don't know how a subclass of `float` that implements its own `__float__` should behave.

```
>>> class MyFloat(float):
...     def __float__(self):
...         return 9.0
... 
>>> x = MyFloat(16)
>>> math.sqrt(x)
4.0
>>> float(x)
9.0
```

File: Modules/mathmodule.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Modules/mathmodule.c]: Good idea!

**Comment by serhiy-storchaka** [comment]
I think that it would be better to not call `__float__` for `float` subclasses, `__int__` and `__index__` for `int` subclasses, `__complex__` for `complex` subclasses, `__str__` for `str` subclasses, etc. We already have a content of corresponding class and can convert it to a C value or to a new Python instance.

**Comment by serhiy-storchaka** [comment]
Thank you for your review @mdickinson, I addressed your comments and added tests. Could you please take another look?

**Pull request merged**